Generate more certs/identities, and use them for better multi-cert test coverage

sam-github opened the pull request:
- test: cover tls multi-identity option mixtures
- test: confirm tls server suite default is its own
- test: add independent multi-alg crypto identities
- test: rename agent1-pfx.pem to agent1.pfx

##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
tls

sam-github commented:
Will rebase onto https://github.com/nodejs/node/pull/10389 once it lands.

sam-github commented:
Rebased onto https://github.com/nodejs/node/pull/10771 so that tests can pass. Will rebase again once #10771 merges.

sam-github commented:
@nodejs/crypto

sam-github commented:
@indutny @nodejs/crypto

bnoordhuis commented on test/fixtures/keys/Makefile:
```diff
@@ -254,6 +374,75 @@ agent7-cert.pem: agent7-csr.pem fake-cnnic-root-cert.pem fake-cnnic-root-key.pem
+	cat ca4-cert.pem >> agent8-cert.pem
+
+agent8-verify: agent8-cert.pem ca4-cert.pem
+	openssl verify -CAfile ca4-cert.pem agent8-cert.pem
```
Target should be .PHONY.

bnoordhuis commented on test/fixtures/keys/Makefile:
```diff
@@ -254,6 +374,75 @@ agent7-cert.pem: agent7-csr.pem fake-cnnic-root-cert.pem fake-cnnic-root-key.pem
+		-inkey ec8-key.pem \
+		-certfile ca6-cert.pem \
+		-out ec8.pfx \
+		-password pass:sample
```
You could make the recipes a little more DRY by using $@, $< and $^ more.

bnoordhuis commented on test/fixtures/keys/ca4.cnf:
```diff
@@ -0,0 +1,23 @@
+O                      = Joyent
+OU                     = Node.js
+CN                     = ca4
+emailAddress           = ry@tinyclouds.org
```
Perhaps switch to something more Node.js Foundation-y.

bnoordhuis replied:
Still relevant.  Not essential, just nice-to-have.

bnoordhuis commented on test/fixtures/keys/ca5.cnf:
```diff
@@ -0,0 +1,31 @@
+O                      = Joyent
+OU                     = Node.js
+CN                     = ca5
+emailAddress           = ry@tinyclouds.org
```
Ditto.

bnoordhuis commented on test/fixtures/keys/ca6.cnf:
```diff
@@ -0,0 +1,22 @@
+O                      = Joyent
+OU                     = Node.js
+CN                     = ca6
+emailAddress           = ry@tinyclouds.org
```
Ditto.

bnoordhuis commented on test/parallel/test-tls-honorcipherorder.js:
```diff
@@ -46,54 +45,50 @@ function test(honorCipherOrder, clientCipher, expectedCipher, cb) {
-      if (cb) cb();
-    });
-  });
+      const msg = fmt(
```
Can you use util.format here?  Saves the casual reader from having to look up what fmt() is.

bnoordhuis commented on test/parallel/test-tls-honorcipherorder.js:
```diff
@@ -1,43 +1,42 @@
-    ciphers: 'AES256-SHA256:AES128-GCM-SHA256:AES128-SHA256:' +
+    ciphers: 'AES256-SHA256:' +
+             'AES128-GCM-SHA256:' +
+             'AES128-SHA256:' +
```
Stylistic change.

silverwind commented:
Are there any matching cert + intermediate + key files in here? Asking so I can test https://github.com/nodejs/node/pull/7230#issuecomment-277475821.

sam-github commented:
@silverwind https://github.com/nodejs/node/pull/10747/commits/5c700f9308e63ad09390b33e4ae8171139d0b6a9

sam-github commented:
@silverwind bf43a60a383fa5fce1051532c4e499c4b1df4c07 has the tests you are looking for, I believe, but this PR is partially broken at the moment, I am in process of rebasing (master had an agent8 and agent9 added to it, so I need to move my numbered certs forward, which I'm in process of doing right now).

silverwind commented:
Hmm, can't find a fitting case there, but this should fail:

```
cert: [agent10-cert, ca4-cert]
key: agent10-key
```

sam-github commented:
@silverwind 

I think I see what is happening, you are trying to provide the certs a+b+c of a cert chain as `cert: [a, b, c]`. This is invalid usage, not documented, and doesn't surprise me it doesn't work (though honestly, it wouldn't surprise me if it had worked, either, ossl is a bit weird). Also, I think you haven't noticed that the PEM files built by the Makefile are *chains* (they have multiple certs in the single PEM file if there are intermediates).

Pre-existing test with pre-existing agent6 cert:

- https://github.com/sam-github/node/blob/d125ee843b015ec9751562945426f620021ba19c/test/parallel/test-tls-multi-key.js#L80-L97

This PR adds tests for mixed algorithm identies, where the mixed algs have intermediate certs:

- https://github.com/nodejs/node/blob/ac602a251a0ca385775a71357d52e9c752430dfb/test/parallel/test-tls-cert-chains-concat.js#L18-L21

Your example:

```
cert: [agent10-cert, ca4-cert]
key: agent10-key
```

is invalid usage. agent10-cert.pem is a *CHAIN*, it already contains ca4-cert.pem, as did the pre-existing cert that had an  intermediate cert, agent6:

- https://github.com/sam-github/node/blob/d125ee843b015ec9751562945426f620021ba19c/test/fixtures/keys/Makefile#L353
- https://github.com/sam-github/node/blob/d125ee843b015ec9751562945426f620021ba19c/test/fixtures/keys/Makefile#L464

Docs say:

> Optional cert chains in PEM format. One cert chain should be provided per private key. Each cert chain should consist of the PEM formatted certificate for a provided private key, followed by the PEM formatted intermediate certificates (if any), in order, and not including the root CA (the root CA must be pre-known to the peer, see ca). 

Your example above has two identities, one is agent10+ca4, and the other is just ca4. This is weird, because ca4 is a sub-chain of the first identity, and also, because they are both RSA, and OpenSSL doesn't permit multiple identities of the same algorithm, because it doesn't know which one to pick (unless you use SNI callback, when it picks by the server name).

sam-github commented:
> So the interesting part is that it works if I concatenate agent10-cert and agent10-int into a single file, but not if I provide them individually in the form [agent10-cert, agent10-int]. I think this should be fixable.

Exactly. That is the documented behaviour, so lets not hijack this PR to discuss it further, now that we agree what is happening. Open another issue if you think there is a problem and we can discuss there.

sam-github commented:
@bnoordhuis I made all requested changes, except the ones that would break consistency. I  think all the certs should be updated together to no longer use Ryan's user names and email, rather than have it be done piecemeal. Its on my todo list. Ditto for the style of the make rules, they could be more concise, but I kept them stylistically identical to the other rules, including the ones just added in https://github.com/nodejs/node/pull/9469

sam-github commented:
@bnoordhuis 

> Are ca3, ca4 and ca6 used anywhere?

ca3 signs agent6 (pre-existing), and agent6 was and is used

ca4 signs agent10, a RSA cert chain with an intermediate, used in the tests in this PR

ca6 signs ec10, a EC cert chain with an intermediate, used in the tests in this PR

sam-github commented:
ci: https://ci.nodejs.org/job/node-test-pull-request/6410/

sam-github commented:
@bnoordhuis PTAL

@nodejs/collaborators needs one more reviewer, its all test code changes, you don't have to be a crypto guru :-)

bnoordhuis commented on test/fixtures/keys/Makefile:
```diff
@@ -1,5 +1,42 @@
+  rsa_private_4096.pem \
+  rsa_public_1024.pem \
+  rsa_public_2048.pem \
+  rsa_public_4096.pem \
```
Since you're here, you might as well sort them in alphabetical order.

sam-github replied:
sorted

bnoordhuis commented on test/fixtures/keys/Makefile:
```diff
@@ -1,5 +1,42 @@
+  rsa_public_4096.pem \
+
+# 0-dns-cert.pem contains an subjectAlternativeName with an embedded NUL, can
+# cannot be created with standard openssl tooling, which is why it is missing
```
"can cannot"

bnoordhuis commented on test/parallel/test-tls-honorcipherorder.js:
```diff
@@ -46,54 +43,50 @@ function test(honorCipherOrder, clientCipher, expectedCipher, cb) {
-  });
+      const msg = util.format(
+        'honorCipherOrder=%j, clientCipher=%j, expect=%j, got=%j',
+        honorCipherOrder, clientCipher, expectedCipher, cipher.name);
```
Can you indent by four spaces?  (I think the linter is supposed to complain.  Not 100% sure though.)

sam-github commented:
ci: https://ci.nodejs.org/job/node-test-pull-request/6446/

sam-github commented:
@nodejs/collaborators need one more review, PTAL, its test-only

shigeki commented:
This leads to an error in re-generating keys as
````
~/github/node/test/fixtures/keys$ make
openssl ca \
        -config fake-startcom-root.cnf \
        -keyfile fake-startcom-root-key.pem \
        -cert fake-startcom-root-cert.pem \
        -batch \
        -days 9999 \
        -passin "pass:password" \
        -in agent8-csr.pem \
        -startdate 20161020235959Z \
        -notext -out agent8-cert.pem
Using configuration from fake-startcom-root.cnf
Error opening CA private key fake-startcom-root-key.pem
140501524727448:error:02001002:system library:fopen:No such file or directory:bss_file.c:398:fopen('fake-startcom-root-key.pem','r')
140501524727448:error:20074002:BIO routines:FILE_CTRL:system lib:bss_file.c:400:
unable to load CA private key
Makefile:407: recipe for target 'agent8-cert.pem' failed
make: *** [agent8-cert.pem] Error 1
````

I'm not sure why it gets conflicts to 89217d145097590e37d11db6e98c317d154be689 which has fake-startcom-root-cert.pem rule in Makefile.

@sam-github Could you rebase this against the latest master?

sam-github commented:
It was already rebased, the conflict occurred because we both added agent8, I changed the name of 8.

I don't really understand what you see, could I have repro instructions?

```
$ make
openssl ca \
        -config fake-startcom-root.cnf \
...
        -notext -out agent8-cert.pem
```

This is odd, since the Makefile doesn't build agent8-cert.pem by default, probably an oversight, so I added this in 6844669. It also didn't depend on its CA private keys

The file mentioned in your error

> Error opening CA private key fake-startcom-root-key.pem

should be a dependency of agent8, I added it in 6844669.

Could you try again?

I can't get the agent8 cert to regenerate on master, though because of a different error, does this work for you?

```
fixtures/keys (master *$% u=) % rm agent8-cert.pem
fixtures/keys (master *$ u=) % make agent8-cert.pem
openssl ca \
        -config fake-startcom-root.cnf \
        -keyfile fake-startcom-root-key.pem \
        -cert fake-startcom-root-cert.pem \
        -batch \
        -days 9999 \
        -passin "pass:password" \
        -in agent8-csr.pem \
        -startdate 20161020235959Z \
        -notext -out agent8-cert.pem
Using configuration from fake-startcom-root.cnf
Check that the request matches the signature
Signature ok
Certificate Details:
        Serial Number: 3 (0x3)
        Validity
            Not Before: Oct 20 23:59:59 2016 GMT
            Not After : Jul  8 21:59:38 2044 GMT
        Subject:
            countryName               = US
            stateOrProvinceName       = CA
            localityName              = SF
            organizationName          = NODEJS
            organizationalUnitName    = agent8
            commonName                = localhost
Certificate is to be certified until Jul  8 21:59:38 2044 GMT (9999 days)
failed to update database
TXT_DB error number 2
Makefile:284: recipe for target 'agent8-cert.pem' failed
make: *** [agent8-cert.pem] Error 1
```

sam-github commented:
@shigeki

shigeki commented:
@sam-github That was my fault in the last PR. The agent9 was also missed.  I will fix it later.

shigeki commented:
The fix was made in 6055401f8a6dd9ecfe7026bbd14f338565ac37c7.  I also found more fixes are needed as in https://github.com/nodejs/node/compare/master...shigeki:fix_regenrate_test_certs. 
I think that opening another new PR is better. Thanks for 6055401f8a6dd9ecfe7026bbd14f338565ac37c7 but it is better to move it to new PR, too for it is not related to the mulit-cert test.

shigeki commented on test/parallel/test-tls-multi-key.js:
```diff
@@ -2,52 +2,162 @@
+});
+
+// Key and cert with mixed algorithms, and cert chains with intermediate CAs
+test({
```
Looking at the cert chain of these tests. They are
```
ec10(agent10):ECC -> ca6:ECC -> ca5(selfsigned):ECC
agent10(agent10):RSA -> ca4:RSA -> ca2(selfsigned):RSA
```
so this cannot formed right cert chain. 
I think we should separate intermidiate cert and root cert as below. 
```diff
@@ -72,6 +76,10 @@ test({
   rsaCN: 'agent10.example.com',
   eccCN: 'agent10.example.com',
   ca: [
+    fs.readFileSync(common.fixturesDir + '/keys/ca4-cert.pem'),
+    fs.readFileSync(common.fixturesDir + '/keys/ca6-cert.pem'),
+  ],
+  rootca: [
     fs.readFileSync(common.fixturesDir + '/keys/ca2-cert.pem'),
     fs.readFileSync(common.fixturesDir + '/keys/ca5-cert.pem'),
   ],
@@ -118,7 +126,7 @@ test({
 function test(options) {
   const rsaCN = options.rsaCN || 'agent1';
   const eccCN = options.eccCN || 'agent2';
-  const ca = options.ca;
+  const rootca = options.rootca;
```

sam-github replied:
@shigeki I don't understand your comment "so this cannot formed right cert chain." Can you elaborate on what is not right?

I don't know if you intentionally said "chain" (singular), because there are two, completely independent cert chains there, the chains you draw out above. They do not share an intermediate cert or a root cert, deliberately, this is exactly the scenario this test is intended to test: two completely independent cert chains. What is not "right"? Also, the tests pass... so the certs can't be that wrong. :-)

I don't understand your suggested change. `.rootca` is not a supported option by node, but perhaps your concern is that the .ca property was being passed to (but ignored) by the server, I removed that to make it more clear that .cert+.key is  used by the server, and .ca is  used by the client. See f6d239e

shigeki replied:
I should have used another proper word. What I worried were that `options.ca` for server you fixed in f6d239e35ddab0950265dad3620593d30aa2b014 which is no use for this test and concatenation intermediate cert of ca4 and ca4 into server certs which results hiding intermediate cert names from the file name. 
This test intends to check two independent cert chains of ECC and RSA works fine so that I think it should be clearly visible what cert chains are testing. Another my proposal is like 
```js
const options = { // [ECC, RSA]
  server: { key: [ec10-key, agent10-key],
            cert: [ec10-cert, agent10-cert],
            ca: [ca6-cert, ca4-cert]},
  client: { ca: [ca5-cert, ca3-cert] }
};
```
with removing `cat ca6-cert.pem >> ec10-cert.pem` and `cat ca4-cert.pem >> agent10-cert.pem` in `Makefile` We can distinguish sever and client options and see what cert chains are going to built in this tls connection.

sam-github replied:
The server does not have `.ca` option, but I modified the new client CA option as you suggest.

shigeki commented on test/parallel/test-tls-multi-key.js:
```diff
@@ -2,52 +2,162 @@
 const fs = require('fs');
 
-const options = {
+// Key and cert are ordered as ec, rsa.
```
This test also has different tiers of cert chain. 
```
ec-cert(agent2):ECC (selfsigned)
agent1(agent1RSA):RSA -> ca1:RSA (selfsigned)
```
I think it is better to change the same tier level of the cert chain if we change this test.

sam-github replied:
I did not change the test, it pre-exists: https://github.com/nodejs/node/blob/4f638f6fd73c0936cfa088ebb1d5c075c9565b85/test/parallel/test-tls-multi-key.js#L12-L21

Having a mixture of a self-signed cert+key, and cert+key that is issued by a CA is possible, and supported by the API (obviously, because the tests passes), and I think its good to test unusual scenarios as well as usual/recommended.

I agree that there should be tests with the same tier level, thats why this PR exists. It adds tests that use independent cert chains, where the cert chains are all multi-tier (have intermediate certs). The test you commented on below is like that, for example. And in order to write those tests, I first had to add generation of multi-tier certs, because there weren't any for ECC.

shigeki replied:
I understand it was the same as the previous. I think this PR is a good time to change this. It is not common situation to have two multi certs that one is self-cert and the other is chained with different cert type(ECC and RSA). And in openssl, x509_verify procedure has different code path in case of self-signed. 
This test intends to check two different cert types of ECC and RSA in the same cert store so that I think it is better to have the same cert chain structure tier for checking only two algorithms.

shigeki commented on test/parallel/test-tls-multi-key.js:
```diff
@@ -2,52 +2,162 @@
+      assert.strictEqual(ecdsa.getPeerCertificate().subject.CN, eccCN);
+      // XXX(sam) certs don't currently include EC key info, so depend on
+      // absence of RSA key info to indicate key is EC.
+      assert(!ecdsa.getPeerCertificate().exponent, 'not cert for an RSA key');
```
We do not worry about to check peerCert of ECC attribute. It is enough to confirm that ECDSA cipher has been established if ECC cert is surely working.  But this check does not have harm.

sam-github commented:
re: https://github.com/nodejs/node/pull/10747#issuecomment-281705748, I popped the unrelated fix, I'll leave it to @shigeki to fix

jasnell commented:
Largely rubber stamp LGTM from me once @shigeki is happy with it.

One thing I would note is that it would be extremely helpful to have some documentation around these to help folks understand how and where to use the keys, how and where they are already used, etc.

shigeki commented:
> One thing I would note is that it would be extremely helpful to have some documentation around these to help folks understand how and where to use the keys, how and where they are already used, etc.

I agree. I also would like to change file names and cert attributes to see its use and purpose for ease.

sam-github commented:
@shigeki PTAL

sam-github commented:
@shigeki any more comments I can address?

sam-github commented:
@shigeki ping

sam-github commented:
@shigeki I have responded to all comments, how can I move this forward?

sam-github commented:
ci: https://ci.nodejs.org/job/node-test-pull-request/9285/

BridgeAR commented:
@sam-github would you be so kind and rebase? I guess it is otherwise ready?

BridgeAR commented:
New CI https://ci.nodejs.org/job/node-test-pull-request/9990/

BridgeAR commented:
@mhdawson @bnoordhuis would you be so kind and reconfirm your LG?

BridgeAR commented:
@sam-github the CI is dark red. Seems like some tests need some additional work.
